src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -970,53 +946,29 @@ private AttributeSyntax ParseAttribute()
-                    while (IsMakingProgress(ref lastTokenPosition))
-                    {
-                        if (this.CurrentToken.Kind == SyntaxKind.CloseParenToken)
-                        {
-                            break;
-                        }
-                        else if (this.CurrentToken.Kind == SyntaxKind.CommaToken || this.IsPossibleAttributeArgument())
-                        {
-                            argNodes.AddSeparator(this.EatToken(SyntaxKind.CommaToken));
-                            argNodes.Add(this.ParseAttributeArgument());
-                        }
-                        else if (this.SkipBadAttributeArgumentTokens(ref openParen, argNodes, SyntaxKind.CommaToken) == PostSkipAction.Abort)
-                        {
-                            break;
-                        }
-                    }
-                }
-                else if (this.SkipBadAttributeArgumentTokens(ref openParen, argNodes, SyntaxKind.IdentifierToken) == PostSkipAction.Continue)
-                {
-                    goto tryAgain;
-                }
-            }
+            var argNodes = this.ParseCommaSeparatedSyntaxList(
+                ref openParen,
+                SyntaxKind.CloseParenToken,
+                static @this => @this.IsPossibleAttributeArgument(),
+                static @this => @this.ParseAttributeArgument(),
+                static (@this, openParen, argNodes, expectedKind, closeKind) => skipBadAttributeArgumentTokens(@this, openParen, argNodes, expectedKind, closeKind),
+                allowTrailingSeparator: false);
 
             return _syntaxFactory.AttributeArgumentList(
                 openParen,
-                _pool.ToListAndFree(argNodes),
+                argNodes,
                 this.EatToken(SyntaxKind.CloseParenToken));
-        }
 
-        private PostSkipAction SkipBadAttributeArgumentTokens(ref SyntaxToken openParen, SeparatedSyntaxListBuilder<AttributeArgumentSyntax> list, SyntaxKind expected)
-        {
-            return this.SkipBadSeparatedListTokensWithExpectedKind(ref openParen, list,
-                p => p.CurrentToken.Kind != SyntaxKind.CommaToken && !p.IsPossibleAttributeArgument(),
-                p => p.CurrentToken.Kind == SyntaxKind.CloseParenToken || p.IsTerminator(),
-                expected);
+            static (SyntaxToken openParen, PostSkipAction action) skipBadAttributeArgumentTokens(
+                LanguageParser @this, SyntaxToken openParen, SeparatedSyntaxListBuilder<AttributeArgumentSyntax> list, SyntaxKind expectedKind, SyntaxKind closeKind)
+            {
+                var action = @this.SkipBadSeparatedListTokensWithExpectedKind(ref openParen, list,
+                    static p => p.CurrentToken.Kind != SyntaxKind.CommaToken && !p.IsPossibleAttributeArgument(),
+                    static (p, closeKind) => p.CurrentToken.Kind == closeKind || p.IsTerminator(),
+                    expectedKind, closeKind);
```

Comment:
moved all these lambdas to be static, which necessitated passing information along as a lambda parameter.